### PR TITLE
Implemented Spring Security for Login/Logout

### PR DESCRIPTION
**Pull request by adrianlibenciuc**

### CODE REVIEW

File: src/main/java/us/icarenow/web/config/MvcConfig.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package us.icarenow.web.config;
+
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.servlet.config.annotation.ViewControllerRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+@Configuration
+public class MvcConfig implements WebMvcConfigurer {
+
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/config/MvcConfig.java]: this is not required, spring mvc takes index.html as default root page.
Please remove this class

File: src/main/java/us/icarenow/web/config/WebSecurityConfig.java [hunk 2]
```diff
@@ -10,56 +10,59 @@
...
+                    .antMatchers("/doctor/**").hasRole("DOCTOR")
+                    .antMatchers("/patient/**").hasRole("PATIENT")
+                    .antMatchers("/admin/**","/h2-console/**").hasRole("ADMIN")
+                    .anyRequest().authenticated()
+                    .and()
                 .formLogin()
-                .loginPage("/login")
-                .permitAll()
-                .and()
+                    .loginPage("/login")
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/config/WebSecurityConfig.java]: Nice, please just re-indent the class

File: src/main/java/us/icarenow/web/config/WebSecurityConfig.java [hunk 3]
```diff
@@ -10,56 +10,59 @@
...
+                    .defaultSuccessUrl("/", true)
+                    .permitAll()
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/config/WebSecurityConfig.java]: Does logout works?

File: src/main/java/us/icarenow/web/controller/LoginController.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+
+@Controller
+public class LoginController {
+
+    @GetMapping("/login")
+    public String loginForm() {
+        return "login";
+    }
+
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/controller/LoginController.java]: remove empty lines

File: src/main/java/us/icarenow/web/controller/UserController.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.servlet.ModelAndView;
+import us.icarenow.web.model.dto.UserCreationForm;
+import us.icarenow.web.service.UserService;
+
+import javax.validation.Valid;
+
+@Controller
+public class UserController {
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/controller/UserController.java]: this controller is not part of your requirements, please remove it

File: src/main/java/us/icarenow/web/model/dto/UserCreationForm.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package us.icarenow.web.model.dto;
+
+import javax.validation.constraints.Email;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.Size;
+
+public class UserCreationForm {
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/model/dto/UserCreationForm.java]: This is not part of your requirements, please remove it.
Radu already solved this part

File: src/main/java/us/icarenow/web/repository/UserRepository.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package us.icarenow.web.repository;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import us.icarenow.web.model.entity.User;
+
+import java.util.Optional;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/repository/UserRepository.java:10]: Nice

File: src/main/java/us/icarenow/web/service/UserService.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    private UserRepository userRepo;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    @Autowired
+    private ModelMapper modelMapper;
+
+    public void createUser(UserCreationForm userDTO) {
+        User user = modelMapper.map(userDTO, User.class);
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/service/UserService.java]: this method is part of sign up, and it was already done by Radu.
Please remove it and ensure Radu is setting the status after you merge this. Please add a comment to Radu's PR about this

File: src/main/java/us/icarenow/web/service/UserService.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        if ("libenciuc.adrian@yahoo.com".equals(userDTO.getEmail())) {
+            user.setRoles(Arrays.asList(new Role("ADMIN")));
+        } else {
+            user.setRoles(Arrays.asList(new Role("DOCTOR")));
+        }
+
+        userRepo.save(user);
+    }
+
+    public List<UserDTO> getAllUsers() {
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/service/UserService.java]: Nice done by using model mapper, but you don't need this mehtod for your requirements

File: src/main/java/us/icarenow/web/service/UserService.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                .collect(Collectors.toList());
+    }
+
+    public UserDTO getUser(long id) {
+        return userRepo.findById(id)
+                .map(user -> modelMapper.map(user, UserDTO.class))
+                .orElseThrow(EntityNotFoundException::new);
+    }
+
+    public void update(UserDTO userDTO) {
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/service/UserService.java]: The same, not needed at this phase. please remove it

File: src/main/java/us/icarenow/web/service/UserService.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import us.icarenow.web.model.entity.Role;
+import us.icarenow.web.model.entity.User;
+import us.icarenow.web.repository.UserRepository;
+
+import javax.persistence.EntityNotFoundException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/service/UserService.java:10]: Nice

File: src/main/resources/application.yml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+server:
+  port: 8083
+  servlet:
+    context-path: /
+
+spring:
```
Comment (andreimladin) [inline src/main/resources/application.yml]: this jpa config should be moved in application-dev.yml

File: src/main/resources/templates/index.html [hunk 13]
```diff
@@ -143,26 +143,28 @@
                                 </a>
                             </li>
                             <li class="menu-item">
-                                <a class="item-link" href="payments.html">
-                                    <!-- <span class="link-icon icofont-pay"></span> -->
-                                    <span class="link-text">      </span>
+                                <a class="item-link" th:href="@{/logout}" data-toggle="modal" data-target="#logoutModal">
```
Comment (andreimladin) [inline src/main/resources/templates/index.html]: Nice, does it work?

File: src/main/resources/templates/index.html [hunk 14]
```diff
@@ -286,6 +288,25 @@
...
+            <div class="modal-header">
+                <h5 class="modal-title" id="exampleModalLabel">Ready to Leave?</h5>
+                <button class="close" type="button" data-dismiss="modal" aria-label="Close">
+                    <span aria-hidden="true">×</span>
+                </button>
+            </div>
+            <div class="modal-body">Select "Logout" below if you are ready to end your current session.</div>
+            <div class="modal-footer">
+                <button class="btn btn-secondary" type="button" data-dismiss="modal">Cancel</button>
+                <a class="btn btn-primary" th:href="@{/logout}">Logout</a>
```
Comment (andreimladin) [inline src/main/resources/templates/index.html]: I see that you have 2 links referring to logout. which one is the good one? or both of them?

File: src/main/resources/templates/signup.html [hunk 15]
```diff
@@ -1,5 +1,5 @@
 <!doctype html>
-<html lang="en">
```
Comment (andreimladin) [inline src/main/resources/templates/signup.html]: This is not part of your requirements. Radu already worked here.
Please remove it

File: src/main/resources/static/assets/css/icofont.css [hunk 16]
```diff
@@ -6022,7 +6022,7 @@
 .icofont-shopping-cart:before {
 	content: "\f0cd";
 }
-.icofont-sign-in:before {
```
Comment (andreimladin) [inline src/main/resources/static/assets/css/icofont.css]: if you change only the name of this css class please revert these css files, hard to maintain and find issues/bugs in css files, so I suggest the leave them as they were and use the all names.
This is applicable to all css files

File: src/main/java/us/icarenow/web/service/UserService.java [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    public List<UserDTO> getAllUsers() {
+        return userRepo.findAll()
+                .stream()
+                .map(user -> modelMapper.map(user, UserDTO.class))
+                .collect(Collectors.toList());
+    }
+
+    public UserDTO getUser(long id) {
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/service/UserService.java]: The same, not needed at this phase. please remove it

File: src/main/java/us/icarenow/web/model/dto/UserDTO.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
+package us.icarenow.web.model.dto;
+
+import us.icarenow.web.model.entity.Role;
+
+import java.util.List;
+
+public class UserDTO {
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/model/dto/UserDTO.java]: Not needed, please remove it

File: src/main/java/us/icarenow/web/model/dto/UserLoginForm.java [hunk 19]
```diff
@@ -0,0 +1,29 @@
+package us.icarenow.web.model.dto;
```
Comment (andreimladin) [inline src/main/java/us/icarenow/web/model/dto/UserLoginForm.java]: rename the package to form

**Review by AlinaSastras — DISMISSED** [review]

File: src/main/resources/templates/sign-up.html [hunk 20]
```diff
@@ -52,9 +52,9 @@ <h2 class="h4 mt-0 mb-1">Sign up</h2>
                     <p class="text-muted">Create your Account</p>
 
                     <form>
-                        <div class="form-group">
-                            <input class="form-control" type="text" placeholder="Name">
-                        </div>
+<!--                        <div class="form-group">-->
```
Comment (andreimladin) [inline src/main/resources/templates/sign-up.html:55]: Please remove this, don't comment code that you wont use it

File: src/main/resources/templates/index.html [hunk 21]
```diff
@@ -143,26 +143,28 @@
...
                             </li>
+                        </ul>
+
+                        <ul sec:authorize="isAnonymous()" class="menu-ul">
                             <li class="menu-item">
-                                <a class="item-link" href="sign-in.html">
-                                    <span class="link-icon icofont-nurse"></span>
-                                    <span class="link-text">Sign in</span>
+                                <a class="item-link" th:href="@{/login}">
+                                    <span class="link-icon icofont-login"></span>
```
Comment (andreimladin) [inline src/main/resources/templates/index.html:156]: icofont-sign-in

**Review by andreimladin — APPROVED** [review]

**Review by baciu-radu — APPROVED** [review]

**Review by AlinaSastras — APPROVED** [review]

**Pull request merged**